### DISCUSSION TITLE
zero feeding "Nulleinspeisung"

### DISCUSSION THREAD

**Discussion by Wilder-Kaiser** [SonnenladenGmbH/APsystems-EZ1-API-HomeAssistant#17] [Ideas]
Hi, 
is it possible to create a "zero feeding" in home assistant using a Shelly Pro 3EM?

like adding in the configuration.yaml something like that:
```yaml
number.apsystems_wr_max_output_power: >
  {% set total_active_power = states('sensor.shellypro3em_xxxxxxxxxxxx_total_active_power') | int %}
  {% if total_active_power > 800 %}
    {{ 800 }}
  {% elif total_active_power >= 0 %}
    {{ max(30, total_active_power) }}
  {% else %}
    {{ max(30, number.apsystems_wr_max_output_power | int - total_active_power) }}
  {% endif %}
```

**Comment by mawoka-myblock** [comment]
I'm not an expert ins this regard and this is honestly also the wrong place to ask for help regarding templates in HA config, but I don't see any reason why this shouldn't work.

**Reply by Wilder-Kaiser** [reply]
Thank you for your honest opinion.
I placed the question here since I don't know any other home assistant integration of the ez1-m inverter.
Apologies if I misplaced it. However, I thought samone else already had that idea and implemented it successfully.
I will ask the HA community as well.

**Reply by mawoka-myblock** [reply]
Sure, no problem! Asking in the HA forum will probably more successful, but feel free to post the solution if you find one, here!

**Reply by Moerphium** [reply]
I had the Same in mind before i got my Set of modules and AP-ez1.
after tinkering around and i habe to admit that i Do Not understand the HA-programming language, i withdrew my plans for Nulleinspeisung. Reasons for that are the following with the most significant reasons first:
- it seems like the change rate of the max_output is not settable too often. I tried adjusting the value every second and IT did Not work. This can also be a Problem on my side. But i figured updating the output Power every 5 seconds is the spot were it works.
- output Power is Not Set immediately after the value change for max_output_power. my AP-ez1 needs just a second or two to reach the set value. 
- my Apartment needs pretty exacly 200W during the day on idle. If i work from home this increases to approximately 350-380W. I adjust the max output now depending on these values.
-

My solution now is a simple automation which runs every 5 seconds but only if the AP-ez1 is availible. It then checks for shelly_total_power_used+apez1_current_total_output. The resulting value is the true current consumption of the Apartment. Then i use the case routine to detect if i ned idle Power -> max output is Set to 250W. If i work from home and use more than 200W but less than 450W - > Set max output to 480W. And if there is need for more than 450W - > Set max output to 800W.

Way easier to maintain and easy to Switch on or off on the Dashboard.

Let me know if you habe a solution that works well for nulleinspeisung. I gave up seeing the adjustment times between setting the value and seeing the numbers adjusting.

**Reply by mawoka-myblock** [reply]
Regrading the update interval: The inverter receives the value immediately, but it takes some time to show up in the ui. This could be improved, but as far as I remember, it still took the inverter 1-2 seconds to reply after setting the value.

**Reply by Wilder-Kaiser** [reply]
Thank you for sharing your experience. That sounds like a clever solution!
Regarding the high latancy of the EZ1-M, I changed the script so it only changes in steps of 10 W:

```
automation:
  - alias: "Adjust Photovoltaic Inverter Power Output"
    trigger:
      - platform: numeric_state
        entity_id: sensor.shellypro3em_xxxx_total_active_power
    action:
      - service: number.set_value
        target:
          entity_id: number.apsystems_wr_max_output_power
        data:
          value: >
            {% set consumption = states('sensor.shellypro3em_xxxx_total_active_power') | float %}
            {% set current_output = states('number.apsystems_wr_max_output_power') | float %}
            {% set min_output = 30 %}
            {% set max_output = 800 %}
            {% set step = 10 %}

            {% set home_consumption = consumption - current_output %}
            {% if home_consumption > 0 %}
              {% set desired_output = ((home_consumption // step) + 1) * step %}
              {% set desired_output = current_output + desired_output %}
            {% elif -9 < home_consumption <= 0 %}
              {% set desired_output = current_output %}
            {% else %}
              {% set desired_output = current_output + ((home_consumption // step) * step) %}
            {% endif %}

            {% set desired_output = [desired_output | round(0, 'floor'), max_output] | min %}
            {% set desired_output = [desired_output, min_output] | max %}

            {{ desired_output }}

```
I will ask the home assistant community before running the script in my HA.

Regarding the latency of the EZ1-M: Does anyone know what kind of antenna the EZ1-M uses? I don't think it is SMA...
I want to rule out a bad Wifi strength.
